File: server/bot/lib/integration/test/chat_interaction_test.cpp
#include <fb/bot/integration/chat_interaction_test.h>
#include <fb/bot/integration/game_controller.h>
#include <fb/model/model.h>

using namespace std::chrono_literals;
using namespace fb::bot::integration;

namespace game_reqs = fb::protocol::game::request;
namespace game_resp = fb::protocol::game::response;

chat_interaction_test::chat_interaction_test(game_bot_controller& controller) :
    bot_integration_test(controller, 5) // Use 5 bots for parallel processing
{ }

async::task<void> chat_interaction_test::on_initialize(game_bot_controller& controller)
{
    co_await this->super::on_initialize(controller);
    co_await this->super::arrange_bots_in_line_formation();
}

async::task<void> chat_interaction_test::on_parallel_scenario_started(uint32_t id)
{
    co_return;
}

async::task<void> chat_interaction_test::on_parallel_scenario_finished(uint32_t id)
{
    auto  bots = this->get_test_bots();
    auto& bot  = bots[id];
    bot->chat("/엔피씨제거");
    co_await this->sleep(1s);
    bot->chat(std::format("Chat interaction scenario {} completed successfully", id));
}

generator<bot_integration_test::scenario_t> chat_interaction_test::on_generate_scenario()
{
    co_yield [this]() -> async::task<bool> {
        co_return co_await this->parallel_scenario_1();
    };
}

std::string chat_interaction_test::name() const
{
    return "Chat Interaction Test";
}

async::task<bool> chat_interaction_test::test_scenario_1(int index)
{
    fb::logger::debug("Starting chat interaction scenario 1 test");

    auto  bots   = this->get_test_bots();
    auto& bot    = bots[index];
    auto  passed = true;

    try
    {
        // Step 1: Bot moves down 1 step
        bot->chat("Moving down 1 step");
        co_await bot->direction(DIRECTION::BOTTOM, DEFAULT_TIMEOUT);
        co_await bot->move(DIRECTION::BOTTOM);

        // Step 2: Create NPC using chat command
        bot->chat("Creating NPC 왈숙네");
        auto npc     = co_await bot->create_npc("왈숙네", DEFAULT_TIMEOUT);
        auto npc_oid = npc.oid;

        co_await bot->move(DIRECTION::TOP);
        co_await bot->direction(DIRECTION::BOTTOM, DEFAULT_TIMEOUT);

        // Step 3: Create 진호박 10개
        bot->chat("Creating 진호박 10개");
        co_await bot->create_item("진호박", 10, DEFAULT_TIMEOUT);

        // Step 4: Sell 진호박 5개 via chat
        bot->chat("Selling 진호박 5개 via chat");
        auto base_price     = game_bot::simple_item("진호박", 0).get_price(this->controller);
        auto expected_money = (base_price / 2) * 5;

        std::ignore = co_await bot->request<game_resp::update_internal>(
            game_reqs::chat(false, "진호박 5개 판다"),
            [expected_money](auto& resp) -> bool {
                // Verify money increase
                return resp.ch_money >= expected_money;
            },
            DEFAULT_TIMEOUT);

        // Step 5: Sell remaining 진호박 5개 via chat
        bot->chat("Selling remaining 진호박 5개 via chat");
        expected_money = (base_price / 2) * 10;

        std::ignore = co_await bot->request<game_resp::update_internal>(
            game_reqs::chat(false, "진호박 다 판다"),
            [expected_money](auto& resp) -> bool {
                return resp.ch_money >= expected_money;
            },
            DEFAULT_TIMEOUT);

        // Step 6: Deposit all money via chat
        bot->chat("Depositing all money via chat");
        std::ignore = co_await bot->request<game_resp::update_internal>(
            game_reqs::chat(false, "돈 다 맡아줘"),
            [](auto& resp) -> bool {
                return resp.ch_money == 0;
            },
            DEFAULT_TIMEOUT);

        // Step 7: Withdraw all money via chat
        bot->chat("Withdrawing all money via chat");
        std::ignore = co_await bot->request<game_resp::update_internal>(
            game_reqs::chat(false, "돈 다 돌려줘"),
            [](auto& resp) -> bool {
                return resp.ch_money > 0;
            },
            DEFAULT_TIMEOUT);

        // Step 8: Get 동동주 via chat
        bot->chat("Getting 동동주 via chat");
        std::ignore = co_await bot->request<game_resp::item_update>(
            game_reqs::chat(false, "동동주 줘"),
            [](auto& resp) -> bool {
                return resp.name.find("동동주") == 0;
            },
            DEFAULT_TIMEOUT);

        // Step 9: Deposit 동동주 via chat
        bot->chat("Depositing 동동주 via chat");
        std::ignore =
            co_await bot->request<game_resp::item_remove>(game_reqs::chat(false, "동동주 맡아줘"), DEFAULT_TIMEOUT);

        // Step 10: Withdraw 동동주 via chat
        bot->chat("Withdrawing 동동주 via chat");
        std::ignore = co_await bot->request<game_resp::item_update>(
            game_reqs::chat(false, "동동주 돌려줘"),
            [](auto& resp) -> bool {
                return resp.name.find("동동주") == 0;
            },
            DEFAULT_TIMEOUT);

        // Step 11: Create 도토리 201개 and deposit all
        bot->chat("Creating 도토리 201개 and depositing all");
        co_await bot->create_item("도토리", 201, DEFAULT_TIMEOUT);

        std::ignore = co_await bot->request<game_resp::item_remove>(
            game_reqs::chat(false, "도토리 다 맡아줘"),
            [](auto& resp) -> bool {
                return true; // Just verify response received
            },
            DEFAULT_TIMEOUT);

        // Step 12: Create 도토리 201개 again and deposit all
        bot->chat("Creating 도토리 201개 again and depositing all");
        co_await bot->create_item("도토리", 201, DEFAULT_TIMEOUT);

        std::ignore = co_await bot->request<game_resp::item_remove>(
            game_reqs::chat(false, "도토리 다 맡아줘"),
            [](auto& resp) -> bool {
                return true; // Just verify response received
            },
            DEFAULT_TIMEOUT);

        // Step 13: Try to withdraw all 도토리 (should fail)
        bot->chat("Trying to withdraw all 도토리 (should fail)");
        std::ignore = co_await bot->request<game_resp::chat>(
            game_reqs::chat(false, "도토리 다 돌려줘"),
            [npc_oid](auto& resp) -> bool {
                return resp.text.find(_TEXT(MESSAGE_ITEM_CANNOT_PICKUP_ANYMORE)) != std::string::npos;
            },
            DEFAULT_TIMEOUT);
    }
    catch (const std::exception& e)
    {
        fb::logger::fatal("Chat interaction scenario 1 failed: {}", e.what());
        passed = false;
    }

    co_return passed;
}

async::task<bool> chat_interaction_test::test_scenario_2(int index)
{
    fb::logger::debug("Starting chat interaction scenario 2 test");

    auto  bots   = this->get_test_bots();
    auto& bot    = bots[index];
    auto  passed = true;

    try
    {
        bot->chat("Moving down 1 step");
        co_await bot->direction(DIRECTION::BOTTOM, DEFAULT_TIMEOUT);
        co_await bot->move(DIRECTION::BOTTOM);

        bot->chat("Creating NPC 떡쇠");
        auto npc     = co_await bot->create_npc("떡쇠", DEFAULT_TIMEOUT);
        auto npc_oid = npc.oid;

        co_await bot->move(DIRECTION::TOP);
        co_await bot->direction(DIRECTION::BOTTOM, DEFAULT_TIMEOUT);

        co_await bot->create_item("양첨목봉", 1, DEFAULT_TIMEOUT);
        co_await bot->change_money(100000, DEFAULT_TIMEOUT);

        std::ignore = co_await bot->request<game_resp::chat>(
            game_reqs::chat(false, "전부 고쳐줘"),
            [npc_oid](auto& resp) -> bool {
                if (resp.oid != npc_oid)
                    return false;

                return resp.text.find(_TEXT(MESSAGE_ITEM_NOT_REPAIRABLE)) != std::string::npos;
            },
            DEFAULT_TIMEOUT);

        bot->chat("/내구도 1");
        co_await this->sleep(1s);

        std::ignore = co_await bot->request<game_resp::chat>(
            game_reqs::chat(false, "전부 고쳐줘"),
            [npc_oid](auto& resp) -> bool {
                if (resp.oid != npc_oid)
                    return false;

                return resp.text.find("고치는데 9900전이 들었습니다.") != std::string::npos;
            },
            DEFAULT_TIMEOUT);

        std::ignore = co_await bot->request<game_resp::item_update>(
            game_reqs::chat(false, "양첨목봉 이름을 채승현으로 명명"),
            [npc_oid](auto& resp) -> bool {
                if (resp.index != 0)
                    return false;

                return resp.name == "채승현";
            },
            DEFAULT_TIMEOUT);
    }
    catch (const std::exception& e)
    {
        fb::logger::fatal("Chat interaction scenario 2 failed: {}", e.what());
        passed = false;
    }

    co_return passed;
}

async::task<bool> chat_interaction_test::test_scenario_3(int index)
{
    auto  bots   = this->get_test_bots();
    auto& bot    = bots[index];
    auto  passed = true;

    try
    {
        bot->chat("Moving down 1 step");
        co_await bot->direction(DIRECTION::BOTTOM, DEFAULT_TIMEOUT);
        co_await bot->move(DIRECTION::BOTTOM);

        bot->chat("Creating NPC 좌성황");
        auto npc     = co_await bot->create_npc("좌성황", DEFAULT_TIMEOUT);
        auto npc_oid = npc.oid;

        co_await bot->move(DIRECTION::TOP);
        co_await bot->direction(DIRECTION::BOTTOM, DEFAULT_TIMEOUT);

        auto spell_index = co_await bot->learn_spell("헬파이어", DEFAULT_TIMEOUT);
        fb::logger::debug("Spell index: {}", spell_index);
        co_await bot->change_mp(100000, DEFAULT_TIMEOUT);

        std::ignore = co_await bot->request<game_resp::update_internal>(
            game_reqs::spell_cast(SPELL_TYPE::TARGET, spell_index, "", bot->oid(), bot->position()),
            [](auto& resp) -> bool {
                return resp.ch_hp == 0;
            },
            DEFAULT_TIMEOUT);

        while (true)
        {
            auto&& resp = co_await bot->request<game_resp::chat>(
                game_reqs::chat(false, "살려주세요"),
                [npc_oid](auto& resp) -> bool {
                    return resp.oid == npc_oid;
                },
                DEFAULT_TIMEOUT);

            if (resp.text.find("너의 정성에 감복하여 살려줄터이니 잠시 기다려라....") != std::string::npos)
                break;
        }

        // Wait for revive
        while (bot->state() != STATE::NORMAL)
        {
            co_await this->sleep(100ms);
        }

        std::ignore = co_await bot->request<game_resp::update_internal>(
            game_reqs::chat(false, "감사합니다"),
            [bot, npc_oid](auto& resp) -> bool {
                return resp.ch_hp == bot->base_hp();
            },
            DEFAULT_TIMEOUT);
    }
    catch (const std::exception& e)
    {
        fb::logger::fatal("Chat interaction scenario 3 failed: {}", e.what());
        passed = false;
    }

    co_return passed;
}

async::task<bool> chat_interaction_test::test_scenario_4(int index)
{
    auto  bots   = this->get_test_bots();
    auto& bot    = bots[index];
    auto  passed = true;

    try
    {
        bot->chat("Moving down 1 step");
        co_await bot->direction(DIRECTION::BOTTOM, DEFAULT_TIMEOUT);
        co_await bot->move(DIRECTION::BOTTOM);

        bot->chat("Creating NPC 뭉치");
        auto npc     = co_await bot->create_npc("뭉치", DEFAULT_TIMEOUT);
        auto npc_oid = npc.oid;

        co_await bot->move(DIRECTION::TOP);
        co_await bot->direction(DIRECTION::BOTTOM, DEFAULT_TIMEOUT);

        std::ignore = co_await bot->request<fb::bot::integration::dialog_ext_bot>(
            game_reqs::click(npc_oid),
            [](auto& resp) -> bool {
                return resp.type == dialog_ext_type::list;
            },
            DEFAULT_TIMEOUT);

        std::ignore = co_await bot->request<fb::bot::integration::dialog_ext_bot>(
            game_reqs::dialog(game_reqs::dialog::INTERACTION::LIST, 0, "", 0, 0, "", DIALOG_RESULT::NEXT),
            [](auto& resp) -> bool {
                return resp.type == dialog_ext_type::list;
            },
            DEFAULT_TIMEOUT);

        std::ignore = co_await bot->request<fb::bot::integration::dialog_bot>(
            game_reqs::dialog(game_reqs::dialog::INTERACTION::LIST, 0, "", 0, 0, "", DIALOG_RESULT::NEXT),
            [](auto& resp) -> bool {
                return resp.type == dialog_type::item;
            },
            DEFAULT_TIMEOUT);

        auto   item_name = "unknown";
        auto&& resp      = co_await bot->request<fb::bot::integration::dialog_ext_bot>(
            game_reqs::dialog(game_reqs::dialog::INTERACTION::ITEM, 0, item_name, 0, 0, "", DIALOG_RESULT::NEXT),
            [](auto& resp) -> bool {
                return resp.type == dialog_ext_type::normal;
            },
            DEFAULT_TIMEOUT);

        passed = resp.message == "This player is a hacker.";
    }
    catch (const std::exception& e)
    {
        fb::logger::fatal("Chat interaction scenario 3 failed: {}", e.what());
        passed = false;
    }

    co_return passed;
}

async::task<bool> chat_interaction_test::parallel_scenario_1()
{
    auto bots      = this->get_test_bots();
    auto scenarios = std::vector<std::pair<uint32_t, scenario_t>>{
        {0,
         [this,    &bots]() -> async::task<bool> {
             co_return co_await this->test_scenario_1(0);
         }   },
        {1,
         [this,    &bots]() -> async::task<bool> {
             co_return co_await this->test_scenario_2(1);
         }   },
        {2,
         [this,    &bots]() -> async::task<bool> {
             co_return co_await this->test_scenario_3(2);
         }   },
        {3, [this, &bots]() -> async::task<bool> {
             co_return co_await this->test_scenario_4(3);
         }}
    };
    co_return co_await this->parallel_scenarios(scenarios);
}

File: scripts/integration/chat_interaction_test.lua
local lib      = require("integration.lib")
local resp     = require("integration.response")
local protocol = require("integration.protocol")

local CANNOT_PICKUP_TEXT = "더 이상 가질 수 없습니다."

local function setup_npc(ctx, bot, index, npc_name)
    log("debug", string.format("Chat[%d]: %s", index, "setup_npc: " .. npc_name))
    bot:direction("BOTTOM")
    bot:move("BOTTOM")

    local npc = bot:create_npc(npc_name)
    log("debug", string.format("Chat[%d]: %s", index, "setup_npc: created npc oid=" .. tostring(npc.oid)))

    bot:move("TOP")
    bot:direction("BOTTOM")

    return npc
end

local function test_scenario_1(ctx, index)
    log("debug", string.format("Chat[%d]: %s", index, "scenario 1 started"))
    local bot = ctx:bot(index)

    setup_npc(ctx, bot, index, "왈숙네")

    bot:create_item("진호박", 10)
    log("debug", string.format("Chat[%d]: %s", index, "scenario 1: created 진호박 x10"))

    local base_price     = bot:item_base_price("진호박")
    local expected_money = math.floor(base_price / 2) * 5
    log("debug", string.format("Chat[%d]: %s", index, "scenario 1: selling 5 pumpkins (expected money >= " .. tostring(expected_money) .. ")"))

    bot:request(
        resp.update_internal,
        protocol.chat(false, "진호박 5개 판다"),
        function(packet)
            return packet.ch_money >= expected_money
        end)
    log("debug", string.format("Chat[%d]: %s", index, "scenario 1: sold 5 pumpkins, money=" .. tostring(bot:money())))
    if bot:money() < expected_money then
        log("fatal", string.format("Chat[%d]: %s", index, "scenario 1 FAILED: money after selling 5"))
        return false
    end

    expected_money = math.floor(base_price / 2) * 10
    log("debug", string.format("Chat[%d]: %s", index, "scenario 1: selling all pumpkins"))
    bot:request(
        resp.update_internal,
        protocol.chat(false, "진호박 다 판다"),
        function(packet)
            return packet.ch_money >= expected_money
        end)
    log("debug", string.format("Chat[%d]: %s", index, "scenario 1: sold all pumpkins, money=" .. tostring(bot:money())))
    if bot:money() < expected_money then
        log("fatal", string.format("Chat[%d]: %s", index, "scenario 1 FAILED: money after selling all"))
        return false
    end

    log("debug", string.format("Chat[%d]: %s", index, "scenario 1: depositing all money"))
    bot:request(
        resp.update_internal,
        protocol.chat(false, "돈 다 맡아줘"),
        function(packet)
            return packet.ch_money == 0
        end)
    log("debug", string.format("Chat[%d]: %s", index, "scenario 1: deposited money, money=" .. tostring(bot:money())))
    if bot:money() ~= 0 then
        log("fatal", string.format("Chat[%d]: %s", index, "scenario 1 FAILED: money not zero after deposit"))
        return false
    end

    log("debug", string.format("Chat[%d]: %s", index, "scenario 1: withdrawing all money"))
    bot:request(
        resp.update_internal,
        protocol.chat(false, "돈 다 돌려줘"),
        function(packet)
            return packet.ch_money > 0
        end)
    log("debug", string.format("Chat[%d]: %s", index, "scenario 1: withdrew money, money=" .. tostring(bot:money())))
    if bot:money() <= 0 then
        log("fatal", string.format("Chat[%d]: %s", index, "scenario 1 FAILED: money not restored"))
        return false
    end

    log("debug", string.format("Chat[%d]: %s", index, "scenario 1: requesting 동동주"))
    bot:request(
        resp.item_update,
        protocol.chat(false, "동동주 줘"),
        function(packet)
            return string.find(packet.name, "동동주", 1, true) == 1
        end)
    log("debug", string.format("Chat[%d]: %s", index, "scenario 1: received 동동주"))

    log("debug", string.format("Chat[%d]: %s", index, "scenario 1: depositing 동동주"))
    bot:request(
        resp.item_remove,
        protocol.chat(false, "동동주 맡아줘"),
        function()
            return true
        end)
    log("debug", string.format("Chat[%d]: %s", index, "scenario 1: deposited 동동주"))

    log("debug", string.format("Chat[%d]: %s", index, "scenario 1: withdrawing 동동주"))
    bot:request(
        resp.item_update,
        protocol.chat(false, "동동주 돌려줘"),
        function(packet)
            return string.find(packet.name, "동동주", 1, true) == 1
        end)
    log("debug", string.format("Chat[%d]: %s", index, "scenario 1: withdrew 동동주"))

    log("debug", string.format("Chat[%d]: %s", index, "scenario 1: depositing 도토리 (first batch)"))
    bot:create_item("도토리", 201)
    bot:request(
        resp.item_remove,
        protocol.chat(false, "도토리 다 맡아줘"),
        function()
            return true
        end)
    log("debug", string.format("Chat[%d]: %s", index, "scenario 1: deposited first 도토리 batch"))

    log("debug", string.format("Chat[%d]: %s", index, "scenario 1: depositing 도토리 (second batch)"))
    bot:create_item("도토리", 201)
    bot:request(
        resp.item_remove,
        protocol.chat(false, "도토리 다 맡아줘"),
        function()
            return true
        end)
    log("debug", string.format("Chat[%d]: %s", index, "scenario 1: deposited second 도토리 batch"))

    log("debug", string.format("Chat[%d]: %s", index, "scenario 1: withdrawing 도토리 (expect cannot pickup)"))
    bot:request(
        resp.chat,
        protocol.chat(false, "도토리 다 돌려줘"),
        function(packet)
            return string.find(packet.text, CANNOT_PICKUP_TEXT, 1, true) ~= nil
        end)
    log("debug", string.format("Chat[%d]: %s", index, "scenario 1 completed successfully"))
    return true
end

local function test_scenario_2(ctx, index)
    log("debug", string.format("Chat[%d]: %s", index, "scenario 2 started"))
    local bot = ctx:bot(index)

    local npc = setup_npc(ctx, bot, index, "떡쇠")

    bot:create_item("양첨목봉", 1)
    bot:money(100000)
    log("debug", string.format("Chat[%d]: %s", index, "scenario 2: created 양첨목봉, money=100000"))

    log("debug", string.format("Chat[%d]: %s", index, "scenario 2: repair with full durability (expect no items)"))
    bot:request(
        resp.chat,
        protocol.chat(false, "전부 고쳐줘"),
        function(packet)
            if packet.oid ~= npc.oid then
                return false
            end
            return string.find(packet.text, "고칠 물건이 없습니다.", 1, true) ~= nil
        end)
    log("debug", string.format("Chat[%d]: %s", index, "scenario 2: got expected no-repair response"))

    log("debug", string.format("Chat[%d]: %s", index, "scenario 2: setting durability to 1"))
    bot:chat("/내구도 1")
    ctx:sleep(1000)

    log("debug", string.format("Chat[%d]: %s", index, "scenario 2: repair with low durability"))
    bot:request(
        resp.chat,
        protocol.chat(false, "전부 고쳐줘"),
        function(packet)
            if packet.oid ~= npc.oid then
                return false
            end
            return string.find(packet.text, "고치는데 9900전이 들었습니다.", 1, true) ~= nil
        end)
    log("debug", string.format("Chat[%d]: %s", index, "scenario 2: repair completed"))

    log("debug", string.format("Chat[%d]: %s", index, "scenario 2: renaming item"))
    bot:request(
        resp.item_update,
        protocol.chat(false, "양첨목봉 이름을 채승현으로 명명"),
        function(packet)
            return packet.index == 0 and packet.name == "채승현"
        end)
    log("debug", string.format("Chat[%d]: %s", index, "scenario 2 completed successfully"))
    return true
end

local function test_scenario_3(ctx, index)
    log("debug", string.format("Chat[%d]: %s", index, "scenario 3 started"))
    local bot = ctx:bot(index)

    local npc = setup_npc(ctx, bot, index, "좌성황")

    if lib.option.disable_pk_protect(bot) == false then
        log("fatal", string.format("Chat[%d]: %s", index, "scenario 3 FAILED: could not disable PK_PROTECT"))
        return false
    end

    local spell_index = bot:learn_spell("헬파이어")
    bot:set_max_hp_mp(100000, 100000)
    bot:mp(100000)
    log("debug", string.format("Chat[%d]: %s", index, "scenario 3: learned 헬파이어, spell_index=" .. tostring(spell_index)))

    local pos = bot:position()
    log("debug", string.format("Chat[%d]: %s", index, "scenario 3: casting 헬파이어 on self"))
    bot:request(
        resp.update_internal,
        protocol.spell_cast("TARGET", spell_index, "", bot:oid(), pos),
        function(packet)
            return packet.ch_hp == 0
        end)
    log("debug", string.format("Chat[%d]: %s", index, "scenario 3: spell cast done, hp=" .. tostring(bot:hp())))

    local revive_attempt = 0
    while true do
        revive_attempt = revive_attempt + 1
        log("debug", string.format("Chat[%d]: %s", index, "scenario 3: revive request attempt " .. tostring(revive_attempt)))
        local packet = bot:request(
            resp.chat,
            protocol.chat(false, "살려주세요"),
            function(p)
                return p.oid == npc.oid
            end)
        log("debug", string.format("Chat[%d]: %s", index, "scenario 3: revive response: " .. tostring(packet.text)))
        if string.find(packet.text, "너의 정성에 감복하여 살려줄터이니 잠시 기다려라....", 1, true) ~= nil then
            break
        end
    end

    log("debug", string.format("Chat[%d]: %s", index, "scenario 3: waiting for NORMAL state (current=" .. tostring(bot:state()) .. ")"))
    while bot:state() ~= "NORMAL" do
        ctx:sleep(100)
    end
    log("debug", string.format("Chat[%d]: %s", index, "scenario 3: state is NORMAL, hp=" .. tostring(bot:hp())))

    log("debug", string.format("Chat[%d]: %s", index, "scenario 3: thanking NPC"))
    bot:request(
        resp.update_internal,
        protocol.chat(false, "감사합니다"),
        function(packet)
            return packet.ch_hp == bot:base_hp()
        end)
    log("debug", string.format("Chat[%d]: %s", index, "scenario 3: thank you done, hp=" .. tostring(bot:hp()) .. " base_hp=" .. tostring(bot:base_hp())))
    if bot:hp() ~= bot:base_hp() then
        log("fatal", string.format("Chat[%d]: %s", index, "scenario 3 FAILED: hp not restored to base"))
        return false
    end

    log("debug", string.format("Chat[%d]: %s", index, "scenario 3 completed successfully"))
    return true
end

local function test_scenario_4(ctx, index)
    log("debug", string.format("Chat[%d]: %s", index, "scenario 4 started"))
    local bot = ctx:bot(index)

    local npc = setup_npc(ctx, bot, index, "뭉치")
    log("debug", string.format("Chat[%d]: %s", index, "scenario 4: npc oid=" .. tostring(npc.oid)))

    log("debug", string.format("Chat[%d]: %s", index, "scenario 4: clicking NPC (expect pursuit dialog)"))
    bot:request_dialog(
        protocol.click(npc.oid),
        function(packet)
            return packet.type == "pursuit"
        end)
    log("debug", string.format("Chat[%d]: %s", index, "scenario 4: click dialog received (pursuit)"))

    log("debug", string.format("Chat[%d]: %s", index, "scenario 4: select 물건 사기 (expect category pursuit)"))
    bot:request_dialog(
        protocol.dialog("PURSUIT", 0, "", 0, 0, "물건 사기"),
        function(packet)
            return packet.type == "pursuit"
        end)
    log("debug", string.format("Chat[%d]: %s", index, "scenario 4: category pursuit done"))

    log("debug", string.format("Chat[%d]: %s", index, "scenario 4: select category (expect item)"))
    bot:request_dialog(
        protocol.dialog("PURSUIT", 0, "", 0, 0, "전사용 갑주류"),
        function(packet)
            return packet.type == "item"
        end)
    log("debug", string.format("Chat[%d]: %s", index, "scenario 4: item dialog done"))

    log("debug", string.format("Chat[%d]: %s", index, "scenario 4: dialog ITEM step (expect normal)"))
    local packet = bot:request_dialog_ext(
        protocol.dialog("ITEM", 0, "", 0, 0, "unknown"),
        function(p)
            return p.type == "normal"
        end)
    log("debug", string.format("Chat[%d]: %s", index, "scenario 4: ITEM step done, message=" .. tostring(packet.message)))

    local ok = packet.message == "This player is a hacker."
    if ok then
        log("debug", string.format("Chat[%d]: %s", index, "scenario 4 completed successfully"))
    else
        log("fatal", string.format("Chat[%d]: %s", index, "scenario 4 FAILED: unexpected message"))
    end
    return ok
end

test_suite {
    name      = "Chat Interaction Test",
    bot_count = 5,

    on_initialize = function(ctx)
        log("debug", "Chat interaction test initialized")
        lib.formation.arrange_in_line(ctx)
    end,

    on_parallel_scenario_finished = function(ctx, id)
        log("debug", "Chat[" .. tostring(id) .. "]: parallel scenario finished, removing NPCs")
        ctx:bot(id):chat("/엔피씨제거")
        ctx:sleep(1000)
    end,

    scenarios = {
        {
            parallel = {
                [0] = { function(ctx) return test_scenario_1(ctx, 0) end },
                [1] = { function(ctx) return test_scenario_2(ctx, 1) end },
                [2] = { function(ctx) return test_scenario_3(ctx, 2) end },
                [3] = { function(ctx) return test_scenario_4(ctx, 3) end },
            },
        },
    },
}
